src/Microsoft.Sbom.Api/PackageDetails/ComponentDetailsUtils/RubyGemsUtils.cs

```diff
@@ -0,0 +1,291 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text.RegularExpressions;
+using Microsoft.ComponentDetection.Contracts.BcdeModels;
+using Microsoft.Sbom.Api.Exceptions;
+using Microsoft.Sbom.Api.Output.Telemetry;
+using Microsoft.Sbom.Common;
+using Serilog;
+using File = System.IO.File;
+
+namespace Microsoft.Sbom.Api.PackageDetails;
+
+/// <summary>
+/// Utilities for retrieving information from rubygems packages that may not be present on the buildDropPath.
+/// </summary>
+public class RubyGemsUtils : IPackageManagerUtils<RubyGemsUtils>
+{
+    private const int ExecutableIndex = 1;
+
+    private readonly IFileSystemUtils fileSystemUtils;
+    private readonly ILogger log;
+    private readonly IRecorder recorder;
+
+    private readonly string rubyGemsPath;
+    private string[] potentialGemEnvPaths;
+
+    public RubyGemsUtils(IFileSystemUtils fileSystemUtils, ILogger log, IRecorder recorder)
+    {
+        this.fileSystemUtils = fileSystemUtils ?? throw new ArgumentNullException(nameof(fileSystemUtils));
+        this.log = log ?? throw new ArgumentNullException(nameof(log));
+        this.recorder = recorder ?? throw new ArgumentNullException(nameof(recorder));
+
+        this.rubyGemsPath = GetRubyGemsSpecificationsPath();
+    }
+
+    // Takes in a scanned component and attempts to find the associated gemspec file. If it is not found then it returns null.
+    public string? GetMetadataLocation(ScannedComponent scannedComponent)
+    {
+        var gemspecLocation = rubyGemsPath;
+
+        if (string.IsNullOrEmpty(gemspecLocation))
+        {
+            return null;
+        }
+
+        var componentName = scannedComponent.Component.PackageUrl?.Name.ToLower();
+        var componentVersion = scannedComponent.Component.PackageUrl?.Version;
+
+        var fullGemspecPath = Path.Join(gemspecLocation, $"{componentName}-{componentVersion}.gemspec");
+
+        if (fileSystemUtils.FileExists(fullGemspecPath))
+        {
+            return fullGemspecPath;
+        }
+        else
+        {
+            // Enumerate directories in the gemspec location and then search for the gemspec file in each directory
+            var directories = fileSystemUtils.GetDirectories(gemspecLocation);
+
+            foreach (var directory in directories)
```

Comment:
Are there any static directory names we ought to ignore or ways we can pare the number of directories we search down?

Reply:
I believe the only nesting that should occur in this scenario is the "default" directory which we could change specifically look for but I figured this was safer in case there are other cases I'm not aware of. But there are none we should ignore.